diff -ruN ruby-1.8.5.orig/ext/socket/socket.c ruby-1.8.5/ext/socket/socket.c
--- ruby-1.8.5.orig/ext/socket/socket.c	2006-08-07 09:50:28.000000000 +0200
+++ ruby-1.8.5/ext/socket/socket.c	2007-02-27 20:43:03.000000000 +0100
@@ -3764,18 +3764,18 @@
     VALUE self, path;
 {
     struct sockaddr_un sockaddr;
-    char *sun_path;
     VALUE addr;
 
     MEMZERO(&sockaddr, struct sockaddr_un, 1);
     sockaddr.sun_family = AF_UNIX;
-    sun_path = StringValueCStr(path);
-    if (sizeof(sockaddr.sun_path) <= strlen(sun_path)) {
+    StringValue(path);
+    if (sizeof(sockaddr.sun_path) <= RSTRING_LEN(path)) {
         rb_raise(rb_eArgError, "too long unix socket path (max: %dbytes)",
             (int)sizeof(sockaddr.sun_path)-1);
     }
-    strncpy(sockaddr.sun_path, sun_path, sizeof(sockaddr.sun_path)-1);
-    addr = rb_str_new((char*)&sockaddr, sizeof(sockaddr));
+    memcpy(sockaddr.sun_path, RSTRING_PTR(path), RSTRING_LEN(path));
+    /* hackish, give me a good spec. */
+    addr = rb_str_new((char*)&sockaddr, RSTRING_LEN(path) + 2);
     OBJ_INFECT(addr, path);
 
     return addr;
